code/renderers/vue3/src/decorateStory.ts

```diff
@@ -14,56 +14,98 @@ import type { VueRenderer } from './types';
+ * @param {DecoratorFunction<VueRenderer>[]} decorators - The array of decorators to apply to the story function.
+ * @returns {LegacyStoryFn<VueRenderer>} - Returns a decorated version of the input storyFn.
+ */
 export function decorateStory(
   storyFn: LegacyStoryFn<VueRenderer>,
   decorators: DecoratorFunction<VueRenderer>[]
 ): LegacyStoryFn<VueRenderer> {
-  return decorators.reduce(
-    (decorated: LegacyStoryFn<VueRenderer>, decorator) => (context: StoryContext<VueRenderer>) => {
-      let story: VueRenderer['storyResult'] | undefined;
+  /**
+   * This function receives two arguments: accuDecoratedStoryFn (the accumulator for the decorated story function) and currentDecoratorFn (the current decorator function).
+   * It applies the decorator to the accuDecoratedStoryFn by returning a new decorated storyFn from the currentDecoratedStory function.
+   *
+   * @param {LegacyStoryFn<VueRenderer>} accuDecoratedStoryFn - The accumulator of the decorated story function.
+   * @param {DecoratorFunction<VueRenderer>} currentDecoratorFn - The current decorator function.
+   */
+  const finalDecoratedStoryFn = decorators.reduce((accuDecoratedStoryFn, currentDecoratorFn) => {
+    let storyResult: VueRenderer['storyResult'];
 
-      const decoratedStory: VueRenderer['storyResult'] = decorator((update) => {
-        const sanitizedUpdate = sanitizeStoryContextUpdate(update);
-        // update the args in a reactive way
-        if (update) sanitizedUpdate.args = Object.assign(context.args, sanitizedUpdate.args);
-        story = decorated({ ...context, ...sanitizedUpdate });
-        return story;
+    /**
+     * This function receives a context argument (the current context of the story function),
+     * and returns a decorated version of the story function.
+     *
+     * @param {StoryContext<VueRenderer>} context - The current context of the story function.
+     */
+    const currentDecoratedStory = (context: StoryContext<VueRenderer>) =>
+      currentDecoratorFn((update) => {
+        const mergedContext = { ...context, ...sanitizeStoryContextUpdate(update) };
+        storyResult = accuDecoratedStoryFn(mergedContext);
+        context.args = mergedContext.args;
+        return storyResult;
       }, context);
 
-      if (!story) story = decorated(context);
-
-      if (decoratedStory === story) {
-        return story;
-      }
+    /**
+     * This function receives a context argument (the current context of the story function),
+     * and returns the final decorated story function with all the decorators applied
+     * @param {StoryContext<VueRenderer>} context - The current context of the story function.
+     * @returns {VueRenderer['storyResult']} - Returns the final decorated story function with all the decorators applied.
+     */
+    return (context: StoryContext<VueRenderer>) => {
+      const story = currentDecoratedStory(context);
+      if (!storyResult) storyResult = accuDecoratedStoryFn(context);
```

Comment:
Doesn't this line mean the later decorators will get called even if the earlier decorator (`currentDecoratorFn`) never calls the passed `storyFn` (defined on line 81)?

As an aside it would help me at least (maybe it's obvious to others) to explain what's happening here -- why do we return `storyResult` sometimes, and other times we call `prepare()` on it?

Reply:
exactly @tmeasday if you don't get a deccorated Story from the decorator or the decorator did you call storyFn, we just take the last one. so we can now have decorator that return null, so we kind of escape them.
so now we can have 
```ts
(storyFn: PartialStoryFn, context: StoryContext) => context.args.condition? storyFn(): null
```

Reply:
> we return `storyResult` sometimes, and other times we call `prepare()` on it?

we don't need to prepare because the previous one is already prepared we just return it

Reply:
we can have decorators that have just side effects and not touching the rendering Tree. by doing something like this.
```ts
(storyFn: PartialStoryFn, context: StoryContext) => {
       saveData(context.args.data)
      return null
  }
```

Reply:
> As an aside it would help me at least (maybe it's obvious to others)

😄 I think you are the only one or at least to first one who spotted the Key change in this PR that fix the conditional decorator issue

Reply:
> we can have decorators that have just side effects and not touching the rendering Tree. by doing something like this.

I'm confused. So a decorator that returns `null` would have "no effect" on rendering, in the sense that later decorators and the story function would still render? That's inconsistent with the other frameworks, and I think the opposite of what we are trying to achieve (consistency in decorator behaviour).

Reply:
no of course the children won't render since the parent not providing any story, one thing is important I'm sure you know it is rendering should happen down -> up the tree. 


https://github.com/storybookjs/storybook/assets/711292/8b4638f0-abd7-45c1-9e78-5bc5f0618260

Reply:
there is a repo on Stackblitz. if you want to test that fast without spawning any sandbox https://stackblitz.com/~/github.com/chakAs3/storybook-vue3-decorators